fuzz/fuzz_lc3.cc

```diff
@@ -0,0 +1,66 @@
+#include "lc3.h"
+#include <array>
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <fuzzer/FuzzedDataProvider.h>
+#include <vector>
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+  FuzzedDataProvider fdp(data, size);
+  auto new_samplerate = [&] {
+    return fdp.PickValueInArray({8000, 16000, 24000, 32000, 48000});
+  };
+
+
+  auto new_pcm_format = [&] {
+    return fdp.PickValueInArray({LC3_PCM_FORMAT_S16, LC3_PCM_FORMAT_S24,
+                                 LC3_PCM_FORMAT_S24_3LE, LC3_PCM_FORMAT_FLOAT});
+  };
+
+  const int dt_us = fdp.PickValueInArray({7500, 10000});
+  const int sr_hz = new_samplerate();
+
+  assert(LC3_CHECK_SR_HZ(sr_hz));
+
+  int sr_pcm_hz = new_samplerate();
+  assert(LC3_CHECK_SR_HZ(sr_pcm_hz));
+
+  if (sr_pcm_hz > sr_hz) {
+    sr_pcm_hz = 0;
+  }
+  const lc3_pcm_format format = new_pcm_format();
+
+  int encoder_size = lc3_encoder_size(dt_us, sr_hz);
+  void *mem = malloc(lc3_encoder_size(dt_us, sr_hz));
+  lc3_encoder_t enc = lc3_setup_encoder(dt_us, sr_hz, sr_pcm_hz, mem);
+
+  if (enc == NULL) {
+    free(mem);
+    return -1;
+  }
+
+  alignas(alignof(uint32_t)) std::array<uint8_t, LC3_MAX_FRAME_BYTES> out = {0};
+  alignas(alignof(uint32_t)) std::array<uint8_t, LC3_MAX_FRAME_BYTES> in = {0};
```

Comment:
So I have a loose understanding of PCM data, but most of this understanding is on the physical (electronics) layer. I'm not sure PCM data looks like at this level of abstraction, or at least there are some gaps in my knowledge (does this represent RZ, NRZ data). Would you say that it's only the floating point data that could be outside of the range of the audio signal or would this apply to (int32_t) data as well? Do you have an easy rule that I could apply that I could use to filter out generated data that wasn't in the audio range?

Reply:
Typically PCM Data is betweem -1.0 to 1.0 in floating point representation. Outside this range, this will lead in saturations (processing saturate within this range).
In Fixed point, it's the same with the comma set just after the sign bit.
It's an open discussion, perhaps makes sense to run with completely out of range values, or special NaN/infinities with floating points.